### PR TITLE
Implement the rest of the struct module

### PR DESCRIPTION
**Pull request by skinny121**

### CODE REVIEW

File: vm/src/byteslike.rs [hunk 1]
```diff
@@ -35,6 +52,44 @@ impl PyBytesLike {
         match self {
             PyBytesLike::Bytes(b) => f(b.get_value()),
             PyBytesLike::Bytearray(b) => f(&b.borrow_value().elements),
+            PyBytesLike::Array(array) => f(&*array.get_bytes()),
+        }
+    }
+}
+
+pub enum PyBuffer {
```
Comment (youknowone) [inline vm/src/byteslike.rs:60]: Oh, buffer protocol is splited from `PyBytesLike`, right?

File: vm/src/stdlib/array.rs [hunk 2]
```diff
@@ -137,14 +140,24 @@ macro_rules! def_array_enum {
                 }
             }
 
-            fn tobytes(&self) -> Vec<u8> {
+            fn get_bytes(&self) -> &[u8] {
```
Comment (youknowone) [inline vm/src/stdlib/array.rs:143]: :+1:

File: vm/src/stdlib/pystruct.rs [hunk 3]
```diff
@@ -160,36 +190,39 @@ mod _struct {
 
     /// Parse endianness
     /// See also: https://docs.python.org/3/library/struct.html?highlight=struct#byte-order-size-and-alignment
-    fn parse_endiannes<I>(chars: &mut Peekable<I>) -> Endianness
+    fn parse_size_and_endiannes<I>(chars: &mut Peekable<I>) -> (bool, Endianness)
```
Comment (youknowone) [inline vm/src/stdlib/pystruct.rs:193]: Probably enum fits better than bool here.

File: vm/src/stdlib/pystruct.rs [hunk 4]
```diff
@@ -72,26 +74,56 @@ mod _struct {
...
-    #[derive(Debug)]
+    #[derive(Debug, Clone)]
     struct FormatSpec {
         endianness: Endianness,
         codes: Vec<FormatCode>,
     }
 
     impl FormatSpec {
+        fn decode_and_parse(
+            vm: &VirtualMachine,
```
Comment (coolreader18) [inline vm/src/stdlib/pystruct.rs:85]: If you'd like, you could probably use `AsciiStr` from the `ascii` crate to simplify parsing/validating fmt strings.

Comment (coolreader18) [reply vm/src/stdlib/pystruct.rs:85]: That would be a lot though, probably best left to a separate PR

File: vm/src/byteslike.rs (hunk 1 above)
Comment (skinny121) [reply vm/src/byteslike.rs:60]: They are pretty much the same except `PyBytesLike` is read only. Looking at https://docs.python.org/3/glossary.html#term-bytes-like-object, this probably should be named `PyRWBytesLike` instead.

File: vm/src/stdlib/pystruct.rs (hunk 4 above)
Comment (skinny121) [reply vm/src/stdlib/pystruct.rs:85]: Yeah will do it in a separate PR.

File: vm/src/stdlib/pystruct.rs [hunk 5]
```diff
@@ -244,6 +286,58 @@ mod _struct {
...
+                let msg = format!(
+                    "{op} requires a buffer of at least {required} bytes for {op_action} {needed} \
+                    bytes at offset {offset} (actual buffer size is {buffer_len})",
+                    op = if is_pack { "pack_into" } else { "unpack_from" },
+                    op_action = if is_pack { "packing" } else { "unpacking" },
+                    required = offset + buffer_len as isize,
+                    needed = needed,
+                    offset = offset,
+                    buffer_len = buffer_len
+                );
```
Comment (youknowone) [inline vm/src/stdlib/pystruct.rs:318]: To clarify `op` and `op_action` is decided by same condition `is_pack`,
```rust
let (op, op_action) = if is_pack {
    ...
} else {
   ...
};
let msg = format!(“...”)
```

File: vm/src/stdlib/pystruct.rs [hunk 6]
```diff
@@ -244,6 +286,58 @@ mod _struct {
...
+                    vm,
+                    format!(
+                        "offset {} out of range for {}-byte buffer",
+                        offset, buffer_len
+                    ),
+                ));
+            }
+            buffer_len - (-offset as usize)
+        } else {
+            if offset as usize >= buffer_len {
```
Comment (youknowone) [inline vm/src/stdlib/pystruct.rs:308]: Unlike then-block, offset is always `usize` in else-block.
```rust
let offset = offset as usize
```

**Pull request merged**